options/options.go

```diff
@@ -1158,6 +1209,95 @@ func (interp *interpreter) setOptionField(mc *internal.MessageContext, msg proto
+// by AST node val. The given name is the AST node that corresponds to the name of fld. On success,
+// it returns additional metadata about the field that was set.
+func (interp *interpreter) setOptionFieldFromProto(mc *internal.MessageContext, msg protoreflect.Message, fld protoreflect.FieldDescriptor, name ast.Node, opt *descriptorpb.UninterpretedOption, node ast.Node) (interpretedFieldValue, error) {
+	k := fld.Kind()
+	var value interpretedFieldValue
+	switch k {
+	case protoreflect.EnumKind:
+		num, _, err := interp.enumFieldValueFromProto(mc, fld.Enum(), opt, node)
+		if err != nil {
+			return interpretedFieldValue{}, err
+		}
+		value = interpretedFieldValue{val: protoreflect.ValueOfEnum(num)}
+
+	case protoreflect.MessageKind, protoreflect.GroupKind:
+		if opt.AggregateValue == nil {
+			return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting message, got %s", mc, optionValueKind(opt))
+		}
+		// We must parse the text format from the aggregate value string
+		fmd := fld.Message()
+		tmpMsg := dynamicpb.NewMessage(fmd)
+		err := prototext.UnmarshalOptions{
+			Resolver:     &msgLiteralResolver{interp: interp, pkg: fld.ParentFile().Package()},
+			AllowPartial: true,
+		}.Unmarshal([]byte(opt.GetAggregateValue()), tmpMsg)
+		if err != nil {
+			return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(node).Start(), "%vfailed to parse message literal %w", mc, err)
+		}
+		msgData, err := proto.MarshalOptions{
+			AllowPartial: true,
+		}.Marshal(tmpMsg)
+		if err != nil {
+			return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(node).Start(), "%vfailed to serialize data from message literal %w", mc, err)
+		}
+		var data []byte
+		if k == protoreflect.GroupKind {
+			data = protowire.AppendTag(data, fld.Number(), protowire.StartGroupType)
+			data = append(data, msgData...)
+			data = protowire.AppendTag(data, fld.Number(), protowire.EndGroupType)
+		} else {
+			data = protowire.AppendTag(nil, fld.Number(), protowire.BytesType)
+			data = protowire.AppendBytes(data, msgData)
+		}
+		// NB: At this point, the serialized fields may no longer be in the same
+		//     order as in the text literal. So for this case, the linker result's
+		//     CanonicalProto won't be in *quite* the right order. ¯\_(ツ)_/¯
+		value = interpretedFieldValue{preserialized: data}
+
+	default:
+		v, err := interp.scalarFieldValueFromProto(mc, descriptorpb.FieldDescriptorProto_Type(k), opt, node)
+		if err != nil {
+			return interpretedFieldValue{}, err
+		}
+		value = interpretedFieldValue{val: protoreflect.ValueOf(v)}
+	}
+
+	if ood := fld.ContainingOneof(); ood != nil {
+		existingFld := msg.WhichOneof(ood)
+		if existingFld != nil && existingFld.Number() != fld.Number() {
+			return interpretedFieldValue{}, reporter.Errorf(interp.nodeInfo(name).Start(), "%voneof %q already has field %q set", mc, ood.Name(), fieldName(existingFld))
```

Comment:
Does protoc error in this case as well?

Reply:
Partially.

```protobuf
syntax = "proto3";

import "google/protobuf/descriptor.proto";

message Foo {
  oneof _1 {
    string frob = 1;
    string nitz = 2;
  }
  oneof _2 {
    string bar = 3;
    string baz = 4;
  }
}

extend google.protobuf.FileOptions {
  Foo foo = 1001;
}

// this generates an error:
//  test.proto:22:16: Error while parsing option value for "foo": Field "nitz" is specified along with field "frob", another member of oneof "_1".
option (foo) = {
  frob: "abc"
  nitz: "def"
};

// but these do not (although they are basically the same as above, but using message de-structuring)
//   see https://github.com/protocolbuffers/protobuf/issues/9125
option (foo).bar = "abc";
option (foo).baz = "def";
```
The comment on the above issue suggests they'd be open to a fix. Anyhow protocompile already behaves this way, when interpreting options from an AST. So this check keeps it consistent, even if it just has a descriptor proto and no AST.